public/app/features/explore/LogsNavigationPages.tsx

```diff
@@ -0,0 +1,123 @@
+import React, { memo } from 'react';
+import { css, cx } from 'emotion';
+import { dateTimeFormat, systemDateFormats, TimeZone, GrafanaTheme, AbsoluteTimeRange } from '@grafana/data';
+import { useTheme, stylesFactory, CustomScrollbar, Spinner } from '@grafana/ui';
+import { LogsPage } from './LogsNavigation';
+
+type Props = {
+  pages: LogsPage[];
+  currentPageIndex: number;
+  oldestLogsFirst: boolean;
+  timeZone: TimeZone;
+  loading: boolean;
+  changeTime: (range: AbsoluteTimeRange) => void;
+};
+
+export function LogsNavigationPages({
+  pages,
+  currentPageIndex,
+  oldestLogsFirst,
+  timeZone,
+  loading,
+  changeTime,
+}: Props) {
+  const formatTime = (time: number) => {
+    return `${dateTimeFormat(time, {
+      format: systemDateFormats.interval.second,
+      timeZone: timeZone,
+    })}`;
+  };
+
+  const createPageContent = (page: LogsPage, index: number) => {
+    if (currentPageIndex === index && loading) {
+      return <Spinner />;
+    }
+    const topContent = formatTime(oldestLogsFirst ? page.logsRange.from : page.logsRange.to);
+    const bottomContent = formatTime(oldestLogsFirst ? page.logsRange.to : page.logsRange.from);
+    return `${topContent} — ${bottomContent}`;
+  };
+
+  const theme = useTheme();
+  const styles = getStyles(theme, loading);
+
+  return (
+    <CustomScrollbar autoHide>
+      <div className={styles.pagesWrapper} data-testid="logsNavigationPages">
+        <div className={styles.pagesContainer}>
+          {pages.map((page: LogsPage, index: number) => (
+            <div
+              data-testid={`page${index + 1}`}
+              className={styles.page}
+              key={page.queryRange.to}
+              onClick={() => !loading && changeTime({ from: page.queryRange.from, to: page.queryRange.to })}
+            >
+              <div className={cx(styles.line, { selectedBg: currentPageIndex === index })} />
+              <div className={cx(styles.time, { selectedText: currentPageIndex === index })}>
+                {createPageContent(page, index)}
+              </div>
+            </div>
+          ))}
+        </div>
+      </div>
+    </CustomScrollbar>
+  );
+}
+
+export default memo(LogsNavigationPages);
```

Comment:
i'm afraid this `memo` will not help much here, the reason is the `changeTime` function coming in the `prop` is recreated in the parent-component on very `render`, so it will be different on every `render`, so the `memo` will rerender.i recommend dropping the `memo` (plan-B is to make `changeTime` not to change on every render, but that does not seem to be easy in this case)

Reply:
Yeah. Another idea would be to memoize the function 
```
  const changeTime = memoizeOne(({ from, to }: AbsoluteTimeRange) => {
    expectedRangeRef.current = { from, to };
    onChangeTime({ from, to });
  });
  ```

But not sure if that is necessary.

Reply:
I have removed the memo.